tools/wake-unit/unit.h

```diff
@@ -0,0 +1,165 @@
+    auto& err = errors.back();
+    err.test_name = test_name;
+    err.file = file;
+    err.line = line;
+    err.predicate_error << "Expected " << term_colour(TERM_MAGENTA) << "`" << cond_str << "`";
+    err.predicate_error << term_normal() << " to be ";
+    err.predicate_error << term_colour(TERM_MAGENTA) << expected_str;
+    err.predicate_error << term_normal() << ", but was found to be ";
+    err.predicate_error << term_colour(TERM_MAGENTA) << actual_str;
+    err.predicate_error << term_normal() << std::endl;
+    return fail(err, assert);
+  }
+
+  TestStream expect_equal(bool assert, std::string expected, std::string actual, const char* expected_str, const char* actual_str, int line, const char* file) {
+    if (expected == actual) return TestStream(nullptr, nullptr);
+    errors.emplace_back();
+    auto& err = errors.back();
+    err.test_name = test_name;
+    err.file = file;
+    err.line = line;
+    err.predicate_error << "Expected:\n\t" << term_colour(TERM_MAGENTA) << quote(expected);
+    err.predicate_error << term_normal() << "\nBut got:\n\t";
+    err.predicate_error << term_colour(TERM_MAGENTA) << quote(actual);
+    err.predicate_error << term_normal() << std::endl;
+    return fail(err, assert);
+  }
+
+  template <class T>
+  TestStream expect_equal(bool assert, T&& expected, T&& actual, const char* expected_str, const char* actual_str, int line, const char* file) {
+    if (expected == actual) return TestStream(nullptr, nullptr);
+    errors.emplace_back();
+    auto& err = errors.back();
+    err.test_name = test_name;
+    err.file = file;
+    err.line = line;
+    err.predicate_error << "Expected " << term_colour(TERM_MAGENTA) << "`" << expected_str << "`";
+    err.predicate_error << term_normal() << " to be equal to `";
+    err.predicate_error << term_colour(TERM_MAGENTA) << actual_str;
+    err.predicate_error << "`" << term_normal() << ", but was found to differ";
+    return fail(err, assert);
+  }
+};
+
+// Public:
+#define EXPECT_TRUE(cond) (logger__.expect(false, true, (cond), #cond, __LINE__, __FILE__))
+#define ASSERT_TRUE(cond) (logger__.expect(true, true, (cond), #cond, __LINE__, __FILE__))
+#define EXPECT_FALSE(cond) (logger__.expect(false, false, (cond), #cond, __LINE__, __FILE__))
+#define ASSERT_FALSE(cond) (logger__.expect(true, false, (cond), #cond, __LINE__, __FILE__))
+#define EXPECT_EQUAL(x, y) (logger__.expect_equal(false, (x), (y), #x, #y, __LINE__, __FILE__))
+#define ASSERT_EQUAL(x, y) (logger__.expect_equal(true, (x), (y), #x, #y, __LINE__, __FILE__))
+
+
+using TestFunc = void(*)(TestLogger&);
+
+struct TestRegister {
+  TestRegister(const char* test_name, TestFunc test);
+};
+
+template <class T>
+struct TestRegisterUnique {
+  static TestRegister unique_register;
+};
+
+#define TEST(name)\
+namespace {struct Test__Unique__ ## name {};}\
+void Test__ ## name (TestLogger&);\
+template<> TestRegister TestRegisterUnique<Test__Unique__ ## name>::unique_register(#name, Test__ ## name);\
```

Comment:
Why all the trickery with templates here? You could have defined a top-level `static` TestRegister.
Are you trying to somehow handle the case where two tests have the same name?

Reply:
This allows two separate files to have the same test name in them because the unique type here is in an anon-namespace. If the same name is used in the same translation unit an error is triggered.

That said I didn't make the test functions themselves static soooo that would be a linker error right now actully. But the goal was to make it so that all errors were local to the translation unit rather than linker errors which this is part of a solution to, I just forgot the easy part.